### PR TITLE
Fix(helmchart): wire options correctly (#7151) and decompose provider

### LINKED ISSUE #7151: helmchart: options wiring gaps (includeCRDs no-op, waitTimeout no-op, install-only gaps for force/maxHistory/cleanupOnFail/recreatePods)
## Summary

Several options declared in the `helmchart` component's `RenderOptionsParams` struct are either complete no-ops or only partially wired. Discovered during validation of the native Helm provider against the documented schema.

### 1. `includeCRDs` is a complete no-op

`IncludeCRDs *bool` exists in `RenderOptionsParams` (helm.go:130) but is never read anywhere in the provider. Neither `newInstallAction` nor the upgrade path references it. Setting `options.includeCRDs: false` has no effect — CRDs from chart templates are always included.

**Expected**: `includeCRDs: false` should exclude CRD resources from the rendered output (maps to Helm SDK's `SkipCRDs` on the install/upgrade action, or filter them from `parseManifestResources`).

### 2. `waitTimeout` is a no-op

`WaitTimeout string` exists in `RenderOptionsParams` (helm.go:138) but is never read. The `timeout` field controls both the rendering timeout and the wait timeout. There is no way for users to set a separate wait timeout.

**Expected**: `waitTimeout` should map to `install.WaitTimeout` / `upgrade.WaitTimeout` on the Helm action when `wait: true` is set.

### 3. `force`, `maxHistory`, `cleanupOnFail`, `recreatePods` are ignored on first install

These four options are correctly wired on the upgrade path (helm.go:1085-1095) but `newInstallAction` does not apply any of them. On a fresh install they are silently ignored.

| Option | Install wired? | Upgrade wired? |
|---|---|---|
| `force` | No | Yes |
| `maxHistory` | No | Yes |
| `cleanupOnFail` | No | Yes |
| `recreatePods` | No | Yes |

For `maxHistory` and `force` in particular, users may set these expecting them to apply from the start. `maxHistory` on install controls the initial history limit; without it the default Helm value is used regardless of what the user set.

## Relevant code

- `pkg/cue/cuex/providers/helm/helm.go`
- `newInstallAction` (line ~1158): missing `IncludeCRDs`, `WaitTimeout`, `Force`, `MaxHistory`, `CleanupOnFail`, `RecreatePods`
- `parseManifestResources` (line ~1448): `IncludeCRDs` filtering not implemented

## Suggested fixes

- `includeCRDs`: filter CRDs in `parseManifestResources` when `IncludeCRDs == false`, or set `install.SkipCRDs` on the action
- `waitTimeout`: apply to `install.WaitTimeout` and `upgrade.WaitTimeout` when `Wait == true`
- `force`, `maxHistory`, `cleanupOnFail`, `recreatePods`: apply in `newInstallAction` with the same logic already used in the upgrade path

## Related

- helmchart component definition: `vela-templates/definitions/internal/component/helmchart.cue`
- Docs PR: https://github.com/kubevela/kubevela.github.io/pull/1433

### PR DESCRIPTION
**Pull request by roguepikachu**
### Description of your changes

Fixes #7151

Companion docs PR: kubevela/kubevela.github.io#1434

**helmchart option-wiring gaps**

- `includeCRDs` is now wired on both install and upgrade. Setting it to `false` correctly skips CRDs from the chart's `crds/` directory.
- `force` is now wired on the install action too; it was already wired on upgrade.
- `waitTimeout` is removed from the schema. The Helm SDK has no separate field for it; the single `timeout` already covers both connection and wait.
- `maxHistory`, `cleanupOnFail`, and `recreatePods` remain upgrade-only by Helm SDK design. The CUE schema, the option appliers, and the user-facing docs now consistently spell out the upgrade-only contract.

**Refactor**

- The helm provider is decomposed into focused per-concern files so future option drift between install and upgrade is visible in code. Install and upgrade each carry their own option applier.
- The unused `Uninstall` provider is removed. The native helmchart ComponentDefinition never invoked it; deletion is handled by KubeVela GC and ResourceTracker.
- Test files mirror the production layout.

I have:

- [x] Read and followed KubeVela's contribution process.
- [x] Related docs updated properly (companion PR on kubevela.github.io).
- [x] Run `make reviewable`.
- [ ] Added `backport release-x.y` labels if needed.

### How has this code been tested

The unit suite remains green throughout the refactor; each refactor step is a single commit gated by `go test`. The final feature commit adds focused tests for the two option appliers and keeps the full suite passing.

A live-cluster validation matrix was executed against this branch covering source detection, option isolation, lifecycle, hooks, self-heal, valuesFrom, and auth. Findings worth documenting (notably a Helm SDK readiness-probe edge case for single-replica deployments) are captured in the companion docs PR.

### Special notes for your reviewer

The refactor commits are pure file moves with no behavior change; tests pass at each step. The user-facing fixes for #7151 land in the final commits. Reviewing commit-by-commit will be easier than reviewing the squashed diff.

### CODE REVIEW

File: pkg/cue/cuex/providers/helm/release.go [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+	// delete it via the ResourceTracker when the Application is deleted.
+	releaseLabels := velaOwnerLabels(velaCtx)
+
+	// Always check the live release in the cluster before using cached data.
+	// This prevents stale cache entries from masking externally-deleted releases
+	// (e.g., helm uninstall, deleted secrets, namespace deletion).
+	getAction := action.NewGet(actionConfig)
+	existingRelease, getErr := getAction.Run(releaseName)
+
+	if getErr != nil {
```
[the remark this answers is not in the document]
Comment (roguepikachu) [reply pkg/cue/cuex/providers/helm/release.go:102]: Addressed. installOrUpgradeChart now only falls through to fresh install when getAction.Run returns driver.ErrReleaseNotFound; any other Get error is wrapped and surfaced so we do not accidentally double-install or wipe the release secrets of a release that actually exists. Thanks for the catch.

File: pkg/cue/cuex/providers/helm/install_test.go [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+kind: ConfigMap
+metadata:
+  name: {{ .Release.Name }}
+`),
+					},
+				},
+			}
+			// Use installOrUpgradeChart which handles actionConfig initialization safely
+			// It will install into a real cluster if available, or fail if not
+			_, _, _, err := p.installOrUpgradeChart(
```
[the remark this answers is not in the document]
Comment (roguepikachu) [reply pkg/cue/cuex/providers/helm/install_test.go]: Addressed by deleting the test. It was a no-cluster smoke test that could perform a real Helm install into default on any machine with cluster access and did not actually assert the failure path it claimed.

File: pkg/cue/cuex/providers/helm/release.go [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+			klog.V(2).Infof("Helm provider [%s]: Release %s held by publishVersion pin %q, skipping upgrade",
+				velaContextStr(velaCtx), releaseName, velaCtx.PublishVersion)
+			p.releaseFingerprints[cacheKey] = fingerprint
+			p.releaseManifests[cacheKey] = existingRelease.Manifest
+			p.releaseVersions[cacheKey] = existingRelease.Version
+			return existingRelease.Manifest, existingRelease.Info.Notes, existingRelease.Version, nil
+		}
+
+		// Release exists — check if it is already deployed with the same fingerprint
+		if !needsAdoption && existingRelease.Info.Status == release.StatusDeployed {
```
[the remark this answers is not in the document]
Comment (roguepikachu) [reply pkg/cue/cuex/providers/helm/release.go]: Addressed. Added the existingRelease.Info != nil guard before reading .Status.

File: pkg/cue/cuex/providers/helm/postrender.go [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+// stored on the Kubernetes release Secret via install.Labels / upgrade.Labels).
+// An external release (installed via `helm install` on the CLI) won't have this label.
+func isOwnedByVela(rel *release.Release, velaCtx *ContextParams) bool {
+	if rel == nil || velaCtx == nil {
+		return false
+	}
+	if rel.Labels == nil {
+		return false
+	}
+	return rel.Labels["app.oam.dev/name"] != ""
```
[the remark this answers is not in the document]
Comment (roguepikachu) [reply pkg/cue/cuex/providers/helm/postrender.go]: Addressed. isOwnedByVela now matches AppName + AppNamespace + Name against the release's stored ownership labels instead of treating any non-empty app.oam.dev/name as ownership. Two Applications targeting the same release name in the same namespace will no longer mis-detect each other as already owned.

File: pkg/cue/cuex/providers/helm/manifest.go [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+// The fallback is intentionally conservative: an unrecognized kind is
+// treated as namespaced, so a new namespaced custom resource gets the
+// safe default (release namespace) rather than landing in vela-system.
+func isClusterScopedGVK(gvk schema.GroupVersionKind) bool {
+	if mapper := singleton.RESTMapper.Get(); mapper != nil {
+		if mapping, mErr := mapper.RESTMapping(gvk.GroupKind(), gvk.Version); mErr == nil && mapping != nil {
+			return mapping.Scope.Name() == meta.RESTScopeNameRoot
+		}
+	}
+	return isClusterScopedKindStaticFallback(gvk.Kind)
```
[the remark this answers is not in the document]
Comment (roguepikachu) [reply pkg/cue/cuex/providers/helm/manifest.go:112]: Acknowledged. The fallback is deliberately conservative: an unrecognized GVK is treated as namespaced, so a CRD-defined cluster-scoped custom resource will be mis-namespaced rather than left in vela-system. The opposite default would skip the namespace-defaulting fix that this code path exists for in the first place, which is a worse failure mode. The function's docstring already spells out the trade-off. A proper fix is two-phase rendering (apply CRDs, refresh RESTMapper, then render the rest); tracked separately from this PR.

File: pkg/cue/cuex/providers/helm/chart_fetch_test.go [hunk 6]
```diff
@@ -0,0 +1,714 @@
...
+			_, _ = w.Write(tgz)
+		}))
+		defer server.Close()
+
+		secret := &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{Name: "creds", Namespace: "rel-ns"},
+			Type:       corev1.SecretTypeBasicAuth,
+			Data:       map[string][]byte{corev1.BasicAuthUsernameKey: []byte("alice"), corev1.BasicAuthPasswordKey: []byte("wonderland")},
+		}
+		singleton.KubeClient.Set(fake.NewClientBuilder().WithScheme(scheme).WithObjects(secret).Build())
```
[the remark this answers is not in the document]
Comment (roguepikachu) [reply pkg/cue/cuex/providers/helm/chart_fetch_test.go:616]: Addressed. fetchOCIChart-with-auth now captures and restores singleton.KubeClient via BeforeEach/AfterEach, matching the sibling fetchURLChart and fetchRepoChart blocks.

File: pkg/cue/cuex/providers/helm/values.go [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+		// CoalesceTables(dst, src) treats dst as authoritative. `values` is the
+		// newer source, so it's passed as dst to override `accumulated` (older).
+		accumulated = chartutil.CoalesceTables(values, accumulated)
+	}
+
+	// Inline values override everything from valuesFrom. Clone before merging
+	// because CoalesceTables mutates dst in place, and dst here is the caller's
+	// map (renderParams.Values).
+	if inline, ok := baseValues.(map[string]interface{}); ok {
+		clone := make(map[string]interface{}, len(inline))
```
[the remark this answers is not in the document]
Comment (roguepikachu) [reply pkg/cue/cuex/providers/helm/values.go]: Addressed. mergeValues now deep-clones inline values via JSON round-trip before CoalesceTables; nested maps can no longer be mutated through to the caller.

File: pkg/cue/cuex/providers/helm/install_test.go [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+				context.Background(), ch, "test-fresh-install", "default",
+				map[string]interface{}{}, nil, nil)
+			// We just verify it doesn't panic - it may succeed or fail depending on cluster
+			_ = err
```
[the remark this answers is not in the document]
Comment (roguepikachu) [reply pkg/cue/cuex/providers/helm/install_test.go]: Addressed by deleting the test (same thread as the not-hermetic finding above). The test asserted only that the call did not panic, which is too weak a check to gate freshInstall regressions.

File: pkg/cue/cuex/providers/helm/values_test.go [hunk 9]
```diff
@@ -0,0 +1,507 @@
...
+			p = NewProviderWithConfig(nil)
+			scheme := runtime.NewScheme()
+			Expect(corev1.AddToScheme(scheme)).To(Succeed())
+			singleton.KubeClient.Set(fake.NewClientBuilder().WithScheme(scheme).Build())
+		})
+
+		It("rejects a ConfigMap reference to a namespace other than the Application's", func() {
+			_, err := p.loadConfigMapValues(context.Background(),
+				ValuesFromParams{Kind: "ConfigMap", Name: "secrets-bearer", Namespace: "kube-system"},
+				"tenant-a", "tenant-a")
```
[the remark this answers is not in the document]
Comment (roguepikachu) [reply pkg/cue/cuex/providers/helm/values_test.go]: Addressed. The cross-namespace guard test now uses distinct app and release namespaces (tenant-a vs release-a) so the assertion exercises both branches of the guard. Also added a positive test for the release-namespace branch.

File: pkg/cue/cuex/providers/helm/dryrun_test.go [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+				Namespace:    "test-ns",
+			}
+			manifest, _, err := p.dryRunRender(ch, "my-rel", "test-ns",
+				map[string]interface{}{}, nil, velaCtx)
+			Expect(err).ShouldNot(HaveOccurred())
+			Expect(manifest).To(ContainSubstring("kind: ConfigMap"))
+			Expect(manifest).To(ContainSubstring("app.oam.dev/name"))
+		})
+
+		It("should apply skipHooks option", func() {
```
[the remark this answers is not in the document]
Comment (roguepikachu) [reply pkg/cue/cuex/providers/helm/dryrun_test.go:101]: Acknowledged. The current spec only verifies that dryRunRender accepts the option without error; verifying that hooks are actually filtered requires a hooks-bearing fixture chart. End-to-end hook handling is exercised against a real chart with hooks in test/e2e-test/helmchart_test.go. Adding an in-memory hook chart purely for this unit test more than doubles the file size; deferring.

File: pkg/cue/cuex/providers/helm/dryrun_test.go [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+			}
+			_, _, err := p.dryRunRender(badChart, "rel", "ns",
+				map[string]interface{}{}, nil, nil)
+			Expect(err).Should(HaveOccurred())
+			Expect(err.Error()).To(ContainSubstring("dry-run render failed"))
+		})
+	})
+
+	Describe("getKubeVersion", func() {
+		It("should not panic and return a result or nil", func() {
```
[the remark this answers is not in the document]
Comment (roguepikachu) [reply pkg/cue/cuex/providers/helm/dryrun_test.go:147]: Acknowledged. Stubbing the discovery client requires either a discovery interface seam in the helm cli.New() pipeline or a custom envtest harness. The current spec only verifies that getKubeVersion does not panic when discovery is unavailable, which is the failure mode that has shown up in practice. Cluster-version round-trip is covered in the e2e suite.

File: pkg/cue/cuex/providers/helm/types.go [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+	SkipHooks       *bool             `json:"skipHooks,omitempty"`
+	CreateNamespace *bool             `json:"createNamespace,omitempty"`
+	Timeout         string            `json:"timeout,omitempty"`
+	MaxHistory      int               `json:"maxHistory,omitempty"`
+	Atomic          bool              `json:"atomic,omitempty"`
+	Wait            bool              `json:"wait,omitempty"`
+	Force           bool              `json:"force,omitempty"`
+	RecreatePods    bool              `json:"recreatePods,omitempty"`
+	CleanupOnFail   bool              `json:"cleanupOnFail,omitempty"`
+	PostRender      *PostRenderParams `json:"postRender,omitempty"`
```
[the remark this answers is not in the document]
Comment (roguepikachu) [reply pkg/cue/cuex/providers/helm/types.go:78]: Acknowledged. PostRenderParams is the wire shape for a planned post-render feature; the field is documented as a future enhancement in the helmchart CUE component definition. Removing the struct now would force a schema break when the feature lands. Tracked alongside the post-render work.

File: pkg/cue/cuex/providers/helm/release_test.go [hunk 13]
```diff
@@ -0,0 +1,154 @@
+/*
```
[the remark this answers is not in the document]
Comment (roguepikachu) [reply pkg/cue/cuex/providers/helm/release_test.go:1]: Addressed. Deleted the two installOrUpgradeChart smoke tests. They could not reach the cluster-dependent code paths they were named for and provided false coverage.

File: pkg/cue/cuex/providers/helm/release_test.go [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+			// empty/optional valuesFrom sources.
+			ch := &chart.Chart{Metadata: &chart.Metadata{Version: "1.2.3"}}
+			fpNil := computeReleaseFingerprint(ch, nil)
+			fpEmpty := computeReleaseFingerprint(ch, map[string]interface{}{})
+			Expect(fpNil).To(Equal(fpEmpty))
+		})
+	})
+
+	Describe("cache invalidation on missing release", func() {
+		It("should not return stale cached data", func() {
```
[the remark this answers is not in the document]
Comment (roguepikachu) [reply pkg/cue/cuex/providers/helm/release_test.go]: Addressed by deleting the test. Same problem: it pre-seeded the cache, called installOrUpgradeChart, and could not actually observe the invalidation branch without a real cluster. A proper assertion lives in the live-cluster validation matrix.

File: pkg/cue/cuex/providers/helm/provider.go [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+			releaseFingerprints: make(map[string]string),
+			releaseManifests:    make(map[string]string),
+			releaseVersions:     make(map[string]int),
+		}
+	})
+	return globalProvider
+}
+
+// NewProviderWithConfig creates a new Helm provider with custom cache configuration
+func NewProviderWithConfig(ttlConfig *CacheTTLConfig) *Provider {
```
[the remark this answers is not in the document]
Comment (roguepikachu) [reply pkg/cue/cuex/providers/helm/provider.go:126]: Acknowledged. utils.NewMemoryCacheStore spawns a background TTL eviction goroutine bound to its constructor context. The singleton provider's worker starts once at process boot via NewProvider() and shares the controller's lifetime. NewProviderWithConfig is used only in tests today and the leak is bounded by the go test process lifetime. A proper fix is to extend MemoryCacheStore with a Stop() method; tracked separately.

File: pkg/cue/cuex/providers/helm/provider.go [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+		MutableVersionTTL:   5 * time.Minute, // 5 minutes for mutable tags
+	}
+}
+
+// Provider is the Helm chart provider
+type Provider struct {
+	cache               *utils.MemoryCacheStore
+	helmClient          *cli.EnvSettings
+	cacheTTL            *CacheTTLConfig
+	releaseMu           sync.Mutex        // serializes install/upgrade/uninstall calls
```
[the remark this answers is not in the document]
Comment (roguepikachu) [reply pkg/cue/cuex/providers/helm/provider.go:84]: Acknowledged. releaseMu serializes install/upgrade across all releases to guard the in-memory fingerprint/manifest/version maps that mutate during each helm action. A per-release sync.Map plus per-key mutex would allow unrelated releases to run concurrently; the change is straightforward but out of scope for #7151. Tracked separately.

File: pkg/cue/cuex/providers/helm/postrender.go [hunk 17]
```diff
@@ -0,0 +1,176 @@
...
+	context          *ContextParams
+	releaseName      string
+	releaseNamespace string
+}
+
+// Run implements postrender.PostRenderer. It parses each YAML document in the
+// rendered manifests, injects KubeVela ownership labels/annotations, and returns
+// the modified manifests.
+func (r *velaLabelPostRenderer) Run(renderedManifests *bytes.Buffer) (*bytes.Buffer, error) {
+	if r.context == nil {
```
[the remark this answers is not in the document]
Comment (roguepikachu) [reply pkg/cue/cuex/providers/helm/postrender.go]: Addressed. velaLabelPostRenderer.Run no longer short-circuits when only the context is nil. Namespace defaulting and Helm ownership annotation injection still run as long as releaseNamespace and releaseName are set, so direct callers benefit from the fix.

File: pkg/cue/cuex/providers/helm/dryrun.go [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+	clientset, err := kubernetes.NewForConfig(cfg)
+	if err != nil {
+		return nil
+	}
+	info, err := clientset.Discovery().ServerVersion()
+	if err != nil {
+		return nil
+	}
+	return &chartutil.KubeVersion{
+		Version: fmt.Sprintf("v%s.%s", info.Major, info.Minor),
```
[the remark this answers is not in the document]
Comment (roguepikachu) [reply pkg/cue/cuex/providers/helm/dryrun.go]: Addressed. getKubeVersion now uses the full GitVersion (e.g. v1.31.4) when present; the major.minor format is the fallback when GitVersion is empty.

File: pkg/cue/cuex/providers/helm/install.go [hunk 19]
```diff
@@ -0,0 +1,121 @@
...
+}
+
+// isRetryableInstallError returns true if the error indicates orphaned state
+// that can be fixed by cleaning up and retrying.
+func isRetryableInstallError(err error) bool {
+	msg := err.Error()
+	return strings.Contains(msg, "cannot be imported") ||
+		strings.Contains(msg, "invalid ownership metadata") ||
+		strings.Contains(msg, "no revision for release") ||
+		strings.Contains(msg, "release: already exists")
```
[the remark this answers is not in the document]
Comment (roguepikachu) [reply pkg/cue/cuex/providers/helm/install.go]: Addressed. isRetryableInstallError now also catches 'cannot reuse a name that is still in use' and the dash-hyphen 'cannot re-use a name that is still in use' variant.

File: pkg/cue/cuex/providers/helm/upgrade.go [hunk 20]
```diff
@@ -0,0 +1,88 @@
...
+
+// applyCommonUpgradeOptions wires user-facing options onto action.Upgrade.
+// MaxHistory, CleanupOnFail, and Recreate are upgrade-only fields per the
+// Helm SDK shape and so live here rather than in applyCommonInstallOptions.
+func applyCommonUpgradeOptions(upgrade *action.Upgrade, opts *RenderOptionsParams) {
+	if opts == nil {
+		return
+	}
+	if opts.IncludeCRDs != nil {
+		upgrade.SkipCRDs = !*opts.IncludeCRDs
```
[the remark this answers is not in the document]
Comment (roguepikachu) [reply pkg/cue/cuex/providers/helm/upgrade.go:67]: Acknowledged. helm v3 only consults SkipCRDs on the install-mode upgrade path (i.e. helm upgrade --install against a release that does not yet exist). Normal upgrades never touch the chart's crds/ directory regardless. I have added an inline comment documenting the limitation so future maintainers do not assume the wiring is operative on every upgrade. A first-class CRD-on-upgrade story would require helm SDK changes; tracked alongside the Helm v4 migration in #7155.

File: pkg/cue/cuex/providers/helm/values.go [hunk 21]
```diff
@@ -0,0 +1,265 @@
...
+// CoalesceTables mutates the destination map (including nested entries) in
+// place, so we must isolate the caller's map at every depth, not just the
+// top level. A JSON round-trip is heavier than `maps.Clone` but is the same
+// technique helm SDK uses internally for the same reason and handles the
+// nested map/slice mix that Helm values can contain.
+func deepCloneValues(in map[string]interface{}) (map[string]interface{}, error) {
+	if in == nil {
+		return nil, nil
+	}
+	buf, err := json.Marshal(in)
```
[the remark this answers is not in the document]
Comment (roguepikachu) [reply pkg/cue/cuex/providers/helm/values.go]: Addressed. Switched from `json.Marshal`/`Unmarshal` to `runtime.DeepCopyJSON`. It walks the same nested map/slice/scalar shape but preserves Go numeric types (int stays int) instead of coercing every number to float64, which would break Helm templates that depend on integer type assertions like `{{ .Values.replicaCount | int }}`.

File: pkg/cue/cuex/providers/helm/values.go [hunk 22]
```diff
@@ -0,0 +1,254 @@
...
+// top level. runtime.DeepCopyJSON is the right tool here: it walks the
+// nested map/slice mix that Helm values can contain AND preserves Go
+// numeric types (int vs float64), where a JSON round-trip would coerce
+// every number to float64 and break Helm templates that rely on integer
+// type assertions (e.g. `{{ .Values.replicaCount | int }}`).
+func deepCloneValues(in map[string]interface{}) map[string]interface{} {
+	if in == nil {
+		return nil
+	}
+	return runtime.DeepCopyJSON(in)
```
[the remark this answers is not in the document]
Comment (roguepikachu) [reply pkg/cue/cuex/providers/helm/values.go]: Addressed. Replaced `runtime.DeepCopyJSON` with a hand-rolled recursive deep-copy. The new helper walks `map[string]interface{}` and `[]interface{}` and returns scalars by interface copy. It accepts the full set of Go numeric types that CUE evaluation can produce (int, int32, int64, float64, ...) instead of panicking on anything that is not a `float64`. Added five focused unit tests that cover int preservation, int64/float64 preservation, nested-map isolation, and slice isolation. Suite passes at 260 specs.

File: pkg/cue/cuex/providers/helm/release_test.go [hunk 23]
```diff
@@ -0,0 +1,277 @@
...
+		Log:          func(format string, v ...interface{}) {},
+	}
+}
+
+// installProviderWithFake returns a fresh Provider whose action-config
+// factory is bound to the supplied fake config, so multiple test specs do
+// not share storage state.
+func installProviderWithFake(cfg *action.Configuration) *Provider {
+	p := NewProviderWithConfig(nil)
+	p.actionConfigFactory = func(string) (*action.Configuration, error) { return cfg, nil }
```
[the remark this answers is not in the document]
Comment (roguepikachu) [reply pkg/cue/cuex/providers/helm/release_test.go:84]: Addressed. Added a second seam on Provider (kubeClientFactory) and routed validateReleaseHealth, listReleaseSecretNames, labelReleaseSecrets, and deleteReleaseSecretsDirect through it. The release_secrets_test specs now build the Provider via a helper that injects clientsetfake.NewSimpleClientset, and the dispatcher tests inject the fake clientset as well. Verified by running the full suite and grepping the output: zero "Kubernetes cluster unreachable" log lines from background goroutines now, so no test path can leak to or mutate the active cluster.

File: pkg/cue/cuex/providers/helm/release_test.go [hunk 24]
```diff
@@ -0,0 +1,277 @@
...
+
+			_, _, version, err := p.installOrUpgradeChart(
+				context.Background(), minimalChart("c", "1.0.0"), relName, relNS,
+				map[string]interface{}{}, nil, velaCtx)
+			Expect(err).ShouldNot(HaveOccurred())
+			Expect(version).To(Equal(2), "adoption should force-upgrade the existing release")
+
+			adopted, err := cfg.Releases.Get(relName, 2)
+			Expect(err).ShouldNot(HaveOccurred())
+			Expect(adopted.Labels).To(HaveKeyWithValue("app.oam.dev/name", "app"))
```
[the remark this answers is not in the document]
Comment (roguepikachu) [reply pkg/cue/cuex/providers/helm/release_test.go:273]: Addressed. The adoption test now asserts all three KubeVela ownership labels (app.oam.dev/name, app.oam.dev/namespace, app.oam.dev/component) on the adopted release, matching the full-context contract enforced by isOwnedByVela.

File: pkg/cue/cuex/providers/helm/release.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (roguepikachu) [reply pkg/cue/cuex/providers/helm/release.go:102]: Update: the strict ErrReleaseNotFound check broke the documented corrupt-secret recovery contract. The e2e regression at `test/e2e-test/helmchart_test.go` ("Helmchart Self-Healing / should recover from corrupted Helm release secret") deliberately corrupts the release secret and expects the dispatcher to auto-recover via the orphan-state cleanup path in freshInstall + isRetryableInstallError. Corrupted secrets surface from the helm storage driver as decode errors, not driver.ErrReleaseNotFound, so the strict check trapped them.

Reverted to fall-through on any non-NotFound error, with a warning log so RBAC / transient API errors are still visible to operators. Added a unit regression that wraps the in-memory storage driver in a corruptingDriver to simulate the e2e scenario without a real cluster; it asserts the dispatcher swallows the corruption error, runs fresh install, and repopulates the cache with the new fingerprint.

The narrower RBAC-protection concern you originally flagged is real but practically narrow: cleanOrphanedReleaseSecrets only runs when the second-stage install fails with one of the isRetryableInstallError patterns, and the same RBAC condition would block its delete calls too. The auto-recovery contract is the binding signal here; the e2e test is the proof.

File: pkg/cue/cuex/providers/helm/release_test.go [hunk 25]
```diff
@@ -0,0 +1,366 @@
...
+			// it must clear the in-memory cache, call freshInstall, and let
+			// the orphan-state retry path clean the corrupted secret.
+			mem := driver.NewMemory()
+			pre := &release.Release{
+				Name: "rel", Namespace: relNS, Version: 1,
+				Info:   &release.Info{Status: release.StatusDeployed},
+				Chart:  minimalChart("c", "1.0.0"),
+				Config: map[string]interface{}{},
+			}
+			Expect(mem.Create("rel.v1", pre)).To(Succeed())
```
[the remark this answers is not in the document]
Comment (roguepikachu) [reply pkg/cue/cuex/providers/helm/release_test.go]: Addressed. Switched the seeded key to the canonical `sh.helm.release.v1.<name>.v<version>` form helm SDK derives via `makeKey`, so the in-memory release record now occupies the same slot the fresh-install path probes. The corruptingDriver layered on top returns a synthetic decode error for every Get/Query targeting that name, forcing the dispatcher down the recovery branch we care about.

Tightened the test comment to be explicit about scope: this unit pins the dispatcher-layer contract (swallow the corruption error + clear the cache + fall through to freshInstall). The end-to-end recovery (delete corrupted Secret via the kube clientset, retry install) is exercised by the e2e test in test/e2e-test/helmchart_test.go ("Helmchart Self-Healing / should recover from corrupted Helm release secret"). The unit gate and the e2e gate together cover the full path.

Also relaxed the fingerprint pre-seed to a realistic-length value so the truncation-safe log path runs cleanly under the test setup. Suite at 267 specs, all green.

**Comment by roguepikachu** [comment]
@cubic-dev-ai review

**Review by anoop2811 — APPROVED** [review]
